COMPACT_DENYLIST_FORMAT.md

```diff
@@ -0,0 +1,458 @@
+(unrelated to the implementation of the parsing/blocking inside IPFS). We
+conciously avoid JSON and other machine formats and opt for text and for
+space-delimited items in a grep/sed/cut-friendly way. For example, we expect
+that the following should just work accross implementations for adding and
+blocking something new:
+
+```
+echo /ipfs/QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768 >> ~/.config/ipfs/custom.deny
+```
+
+We conciously avoid defining any other API other than expecting
+implementations to honor blocking what is on the denylist and act accordingly
+when it is updated. CLI commands or API endpoint to modify list items etc. are
+outside the scope of this spec. Implementations how much information to
+provide to users when a request for an IPFS object is blocked.
+
+As a last note, if we take Kubo and the go-ipfs stack as the reference IPFS
+implementation, we expect the blocking-layer (that is, the introduction of the
+logic that decides whether an item is blocked or not), to happen cleanly at
+the `NameSystem`, `path.Resolver` and `BlockService` interfaces (IPNS, IPFS
+Path and CID blocks respectively).
+
+This specification corresponds to V1 of the compact list format. We have
+limited the number of features and extensions to a minimum to start working
+with, leaving some ideas on the table and the door open to develop the format
+in future versions.
+
+## Specification
+
+### Denylist file extension, locations and order
+
+While not pertaining to the denylist format itself, we introduce the following conventions about denylist files when they are stored in the local filesystem:
+
+- Denylist files are named with the extension `.deny`.
+- Implementations should look in `/etc/ipfs/denylists/` and
+  `$XDG_CONFIG_HOME/ipfs/denylists/` for denylist files.
+- Denylist files are processed in alphabetical order so that rules from later
+  denylists override rules from earlier denylists on conflict.
+
+### Denylist format
+
+#### Summary
+
+The following example showcases the features and syntax of a compact denylist:
+
+```
+version: 1
+name: IPFSCorp blocking list
+description: A collection of bad things we have found in the universe
+author: abuse-ipfscorp@example.com
+hints:
+  hint: value
+  hint2: value2
+---
+# Blocking by CID - blocks wrapped multihash.
+# Does not block subpaths.
+/ipfs/bafybeihvvulpp4evxj7x7armbqcyg6uezzuig6jp3lktpbovlqfkuqeuoq
+
+# Block all subpaths
+/ipfs/QmdWFA9FL52hx3j9EJZPQP1ZUH8Ygi5tLCX2cRDs6knSf8/*
+
+# Block some subpaths (equivalent rules)
+/ipfs/Qmah2YDTfrox4watLCr3YgKyBwvjq8FJZEFdWY6WtJ3Xt2/test*
+/ipfs/QmTuvSQbEDR3sarFAN9kAeXBpiBCyYYNxdxciazBba11eC/test/*
+
+# Block some subpaths with exceptions
+/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked*
++/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blockednot
++/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/not
++/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/exceptions*
+
```

Comment:
```suggestion

# Last-wins resolution of conflicts: 

/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/

!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/allowedsubset

/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/allowedsubset/disallowedsubset

```
(note, I used `!` instead of `+` as negation prefix!)

Reply:
I'm not sure this is correct, the last line should probably be `/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/*` as otherwise it won't match `/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/allowedsubpath` and the path will be allowed (if I'm not mistaken).

I added:

```
# Block some subpaths with exceptions: last-matching-rule wins (!)
```

I think it is illustrative for a showcase of the syntax.